### ISSUE TITLE
Any way to run multiple clients concurrently?

### ISSUE THREAD

**Issue by Bryan-Orifice** [resemble-ai/chatterbox#244]
I want to reduce the latency as much as possible and increase the maximum number of users it can support concurrently on one RTX 5090.  The model fits in VRAM almost 8 times but running multiple workers slows the generation times way down such that it's not helpful (at least the way I'm doing it).  

I've tried chunking and streaming but the generated text to real time playback time is already close to 1:1 with one user. 

Ai suggested that I might be able to "export to TensorRT and use multiple execution contexts" but couldn't confirm if this was possible with chatterbox, so I haven't perused it, not knowing much about this.

Thanks for the great model, rivals the SOTA paid services

**Comment by rsxdalv** [comment]
For non-multilingual, try Chatterbox-VLLM (throughput) or my fork chatterbox fast (focused on latency)

**Comment by jmdBB** [comment]
Hey, you can also load some instances, the GPU load on 3090 can handle at least two concurrent chatterbox instances, it is what Im doing. Im testing more anyways so will see.. Im going to export to TensorRT also and see if I can make it possible.

**Comment by jesulo** [comment]
Did you manage to do it?

**Comment by MansfieldPlumbing** [comment]
@jmdBB @Bryan-Orifice — I actually experimented with the TensorRT route back in March.

I got S3Gen far enough to build and execute as a static-shape TensorRT engine on TRT 10.15 / RTX 3090. It wasn't semantically correct yet: it produced an initial paralinguistic/audio transient and then silence. I moved on to trying dynamic token lengths, and the dynamic-shape/shape-propagation side became the blocker before I got back to it.

The main thing that got TensorRT past S3Gen wasn't a plugin. I stopped trying to preserve the spectral ops at the ONNX boundary.

Chatterbox's HiFT generator has roughly:

```python
spec = torch.stft(..., return_complex=True)
spec = torch.view_as_real(spec)
...
wav = torch.istft(torch.complex(real, imag), ...)
```

I lowered that into real-valued primitive math instead. For the forward transform, the basic approach was:

```python
def decomposed_stft(x, n_fft, hop, window):
    k = torch.arange(
        n_fft // 2 + 1,
        device=x.device,
        dtype=x.dtype
    ).unsqueeze(1)

    n = torch.arange(
        n_fft,
        device=x.device,
        dtype=x.dtype
    ).unsqueeze(0)

    phase = 2.0 * torch.pi * k * n / n_fft

    real = torch.cos(phase) * window.unsqueeze(0)
    imag = -torch.sin(phase) * window.unsqueeze(0)

    weights = torch.cat([real, imag], dim=0).unsqueeze(1)
    return F.conv1d(x.unsqueeze(1), weights, stride=hop)
```

I did the inverse with `conv_transpose1d` + overlap normalization, represented complex values as ordinary real tensors rather than complex dtype, and flattened/restored leading dimensions to keep rank-dependent control flow out of the exported graph.

The resulting ONNX was ugly, but that was intentional: primitive Conv/Reshape/Slice/arithmetic was substantially easier for TensorRT to consume than preserving the high-level spectral semantics.

So the answer to “can Chatterbox/S3Gen get into TensorRT?” is at least **yes, partially**. I got through ONNX parsing, engine construction and execution.

The unfinished problems in my experiment were correct variable-length execution and output semantics, not simply “TensorRT can't parse S3Gen.”

My 3090 died before I came back to it, but I still have the exporter/C++ runtime/notes if anyone actively working on this wants the ugly details.
